extensions/sources/sql/sql.go

```diff
@@ -0,0 +1,139 @@
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package main
+
+import (
+	"database/sql"
+	"fmt"
+	_ "github.com/lf-edge/ekuiper/extensions/sqldatabase/apply"
+	driver2 "github.com/lf-edge/ekuiper/extensions/sqldatabase/driver"
+	"github.com/lf-edge/ekuiper/pkg/api"
+	"github.com/lf-edge/ekuiper/pkg/cast"
+	"github.com/xo/dburl"
+	"time"
+)
+
+type sqlConConfig struct {
+	Interval int    `json:"interval"`
+	Url      string `json:"url"`
+}
+
+type sqlsource struct {
+	conf  *sqlConConfig
+	Query driver2.SqlQueryGenerator
+	//The db connection instance
+	db *sql.DB
+}
+
+func (m *sqlsource) Configure(_ string, props map[string]interface{}) error {
+	cfg := &sqlConConfig{}
+
+	err := cast.MapToStruct(props, cfg)
+	if err != nil {
+		return fmt.Errorf("read properties %v fail with error: %v", props, err)
+	}
+	if cfg.Url == "" {
+		return fmt.Errorf("property Url is required")
+	}
+	if cfg.Interval == 0 {
+		return fmt.Errorf("property interval is required")
+	}
+
+	Db, err := dburl.Parse(cfg.Url)
+	if err != nil {
+		return fmt.Errorf("dburl.Parse %s fail with error: %v", cfg.Url, err)
+	}
+
+	generator, err := driver2.GetQueryGenerator(Db, props)
+	if err != nil {
+		return fmt.Errorf("GetQueryGenerator %s fail with error: %v", cfg.Url, err)
+	}
+
+	m.Query = generator
+	m.conf = cfg
+	return nil
+}
+
+func (m *sqlsource) Open(ctx api.StreamContext, consumer chan<- api.SourceTuple, errCh chan<- error) {
+	logger := ctx.GetLogger()
+	logger.Debugf("Opening sql stream %v", m.conf)
+
+	db, err := dburl.Open(m.conf.Url)
+	if err != nil {
+		logger.Errorf("connection to %s Open with error %v", m.conf.Url, err)
+		errCh <- err
+		return
+	}
+	m.db = db
+
+	t := time.NewTicker(time.Duration(m.conf.Interval) * time.Millisecond)
+	defer t.Stop()
+	for {
+		select {
+		case <-t.C:
+			query, err := m.Query.SqlQueryStatement()
+			if err != nil {
+				logger.Errorf("Get sql query error %v", err)
+			}
+			logger.Debugf("Query the database with %s", query)
+			rows, err := m.db.Query(query)
+			if err != nil {
+				logger.Errorf("Run sql query(%s) error %v", query, err)
+				errCh <- err
+				return
+			}
+
+			cols, _ := rows.Columns()
+			for rows.Next() {
+				data := make(map[string]interface{})
+				columns := make([]interface{}, len(cols))
+				columnPointers := make([]interface{}, len(cols))
+				for i, _ := range columns {
+					columnPointers[i] = &columns[i]
+				}
+
+				err := rows.Scan(columnPointers...)
+				if err != nil {
+					logger.Errorf("Run sql scan(%s) error %v", query, err)
+					errCh <- err
+					return
+				}
+
+				for i, colName := range cols {
+					data[colName] = columns[i]
+				}
+				m.Query.CollectIndexValue(data)
```

Comment:
looks like over wrapping for me. Maybe just count the max value here

Reply:
ack